srt-protocol/src/options/encryption.rs

```diff
@@ -0,0 +1,279 @@
+    ///
+    /// At SRTO_KMPREANNOUNCE packets before switchover the new key is sent (repeatedly, if
+    /// necessary, until it is confirmed by the receiver).
+    ///
+    /// At the switchover point (see SRTO_KMREFRESHRATE), the sender starts encrypting and sending
+    /// packets using the new key. The old key persists in case it is needed to decrypt packets that
+    /// were in the flight window, or retransmitted packets.
+    ///
+    /// The old key is decommissioned at SRTO_KMPREANNOUNCE packets after switchover.
+    ///
+    /// The allowed range for this value is between 1 and half of the current value of
+    /// SRTO_KMREFRESHRATE. The minimum value should never be less than the flight window SRTO_FC
+    /// (i.e. the number of packets that have already left the sender but have not yet arrived at
+    /// the receiver).
+    ///
+    /// The value of SRTO_KMPREANNOUNCE must not exceed (SRTO_KMREFRESHRATE - 1) / 2`.
+    ///
+    /// Default value: 2^12
+    pub pre_announcement_period: usize,
+}
+
+impl Default for KeyMaterialRefresh {
+    fn default() -> Self {
+        Self {
+            pre_announcement_period: 1 << 12, // 2^12,
+            period: 1 << 24,                  // 2^25
+        }
+    }
+}
+
+impl Validation for Encryption {
+    type Error = OptionsError;
+
+    fn is_valid(&self) -> Result<(), Self::Error> {
+        let period = self.km_refresh.period;
+        let pre_announcement_period = self.km_refresh.pre_announcement_period;
+
+        if period == 0 || pre_announcement_period > period.saturating_sub(1) / 2 {
+            Err(OptionsError::KeyMaterialRefresh(
+                period,
+                pre_announcement_period,
+            ))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+// https://github.com/Haivision/srt/blob/master/docs/API/API-socket-options.md#srto_passphrase
+#[derive(Clone, Eq, PartialEq)]
+pub struct Passphrase(String);
+
+impl From<&'static str> for Passphrase {
+    fn from(value: &'static str) -> Self {
+        Self::try_from(value.to_string()).unwrap()
+    }
+}
+
+impl TryFrom<String> for Passphrase {
+    type Error = OptionsError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        if !(10..=79).contains(&value.len()) {
+            return Err(OptionsError::PassphraseLength(value.len()));
+        }
+        Ok(Passphrase(value))
+    }
+}
+
+impl Passphrase {
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+}
+
+impl Display for Passphrase {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        Debug::fmt(self, f)
+    }
+}
+
+impl Debug for Passphrase {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Passphrase").finish()
+    }
+}
+
+// https://github.com/Haivision/srt/blob/master/docs/API/API-socket-options.md#srto_pbkeylen
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum KeySize {
+    Unspecified,
```

Comment:
Why do we need Unspecified?

Reply:
I put this in after noticing in the reference implementation's configuration documentation stated that an unspecified size has some meaning for  how encryption settings get negotiated, so I started to go down the path of implementing that, but I believe there are probably better ways to represent this in the configuration now, so I'll likely pull this out before we merge.

Reply:
What about using an Option<KeySize>?

Reply:
Unspecified is the default, and there is default key size if one is unspecified. Where there are semantics in the domain beyond Some/None, I prefer to encode it in the type, e.g. I can implement Default, and as_usize() and the semantics for a default size will be accessible anywhere. Domain modeling in a type system is often a fight against the urge/bias towards "primitive obsession". Option<T> is a primitive, though surely way better than C/C++ *T